pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,281 @@
+            name = self._scene_cfg[sname]['image_name']
+            self.nout[name] = self._restore_nout(name) if intg.isrestart else 0
+
+        # Set subdivision
+        self.divisor = self.cfg.getint(cfgsect, 'devision', 
+                                       self.cfg.getint('solver', 'order'))
+        self.etypes_div = defaultdict(lambda: self.divisor)
+
+        # Share number of ele types per part
+        ndoms = len(system.ele_map.items())
+        ndoms = comm.gather(ndoms, root=root)
+        if rank == root:
+            ndoms = np.asarray(ndoms)
+        ndoms = comm.bcast(ndoms, root=root)
+        self.dom_0 = dom_0 = np.sum(ndoms[:rank])
+
+        # Generate conduit blueprint of mesh
+        mesh_n = ConduitNode()
+        self.ops = {}
+        for (i, (ename, eles)) in enumerate(system.ele_map.items()):
+            if ename in ['pri', 'pyr']:
+                raise ValueError(f'Ascent: {ename} elements not supported')
+
+            dpf = f'domain_{dom_0 + i}/'
+
+            svpts = self._get_std_ele(ename, eles.nspts)
+            nsvpts = len(svpts)
+
+            mesh_op = self._get_mesh_op(ename, eles.nspts, svpts)
+            soln_op = self._get_soln_op(ename, eles.nspts, svpts)
+            self.ops[ename] = (mesh_op, soln_op)
+
+            mesh_n[dpf + 'state/domain_id'] = rank
+            mesh_n[dpf + 'coordsets/coords/type'] = 'explicit'
+            mesh_n[dpf + 'topologies/mesh/coordset'] = 'coords'
+            mesh_n[dpf + 'topologies/mesh/type'] = 'unstructured'
+            mesh_n[dpf + 'topologies/mesh/elements/shape'] = ename
+
+            subdvcls = subclass_where(BaseShapeSubDiv, name=ename)
+            sconn = subdvcls.subnodes(self.divisor)
+            subdvcon = np.hstack([sconn + j*nsvpts for j in range(eles.neles)])
+            mesh_n[dpf + 'topologies/mesh/elements/connectivity'] = subdvcon
+
+            xd = np.einsum('ik,k...', mesh_op, eles.ploc_at_np('linspts'))
+            X = np.reshape(xd, (eles.ndims, eles.neles*nsvpts))
+            for j in range(eles.ndims):
+                mesh_n[dpf + f"coordsets/coords/values/{('x','y','z')[j]}"] = X[j,:]
+
+            for field in self._expr:
+                mesh_n[dpf + f'fields/{field}/association'] = 'vertex'
+                mesh_n[dpf + f'fields/{field}/volume_dependent'] = 'false'
+                mesh_n[dpf + f'fields/{field}/topology'] = 'mesh'
+
+        self.mesh_n = mesh_n
+
+        # Inital Ascent setup
+        self.a = Ascent()
+        self.ascent_config = ConduitNode()
+        self.ascent_config['mpi_comm'] = comm.py2f()
+        self.ascent_config['runtime/type'] = 'ascent'
+        self.ascent_config['runtine/vtkm/backend'] = 'serial'
+
+        # Open ascent
+        self.a.open(self.ascent_config)
+
+    def __del__(self):
+        comm, rank, root = get_comm_rank_root()
+        # this is needed as if you close while ranks are still 
+        # rendering/writing you might corrutp the file
+        comm.Barrier()
```

Comment:
Not safe; different ranks can call `__del__` in different orders so the barrier may deadlock.

Reply:
What is a better alternative?

Reply:
Ensure you have a barrier before leaving `__call__` so the above scenario of calling close while ranks are writing can not happen.